compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -1648,6 +1648,12 @@ pub enum StatementKind<'tcx> {
     /// This writes `uninit` bytes to the entire place.
     Deinit(Box<Place<'tcx>>),
 
+    /// Marks the place as fully initialized.
+    ///
+    /// This behaves equivalent to an assignment assigning to the place, but allows assigning
+    /// the components of the place incrementally.
+    Finalize(Box<Place<'tcx>>),
```

Comment:
Just to clarify, you mean that this acts like a self-assignment, eg `x = x;`, right? These are not the semantics I would have expected. Originally, I was hoping for this to be a semantic nop and be an "analysis only" artifact (like `FalseRead`) but that does seem non-ideal. What is the motivation for not making this only a read though?

Reply:
Oh, drop elaboration wouldn't know that the place is initialized. Hmm... I do still feel like the part about this being a store is redundant with the `Deinit`, since that already contains sufficient information to mark the place dead. I suppose maybe I'm suggesting that we just teach drop elaboration/borrowck that this means the place was initialized? I don't know, none of these options seem super appealing

Reply:
I changed it (including a force pushed change to `move_paths`, which is used in drop elab), so that we don't treat is as a move, but just mark it as deeply initialized (and `Deinit` marks it as shallowly initialized)